### PR TITLE
chore(warehouse): new event type extract

### PR DESCRIPTION
**Pull request by debanjan97**
## Description of the change

Adds support for a new event type `extract`
which will be used by extract/ETL to send events.

This event type doesn't have the restriction of
sending the version of sources for
mapping `recordId` to `id` column in the warehouse.

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/warehouse/index.js [hunk 1]
```diff
@@ -567,6 +569,80 @@ function processWarehouseMessage(message, options) {
 
   // store columnTypes as each column is set, so as not to call getDataType again
   switch (eventType) {
+    case 'extract': {
+      // set properties common to both tracks and event table
```
Comment (achettyiitr) [inline src/warehouse/index.js:573]: Can we remove this?

File: src/warehouse/index.js [hunk 2]
```diff
@@ -567,6 +569,80 @@ function processWarehouseMessage(message, options) {
...
+        commonColumnTypes,
+        options,
+        'context_',
+      );
+
+      // set event column based on event_text in the tracks table
+      const eventColName = utils.safeColumnName(options, 'event');
+      commonProps[eventColName] = utils.transformTableName(
+        options,
+        message[utils.safeColumnName(options, 'event')],
```
Comment (achettyiitr) [inline src/warehouse/index.js:591]:
```suggestion
        message[eventColName],
```

File: src/warehouse/index.js [hunk 3]
```diff
@@ -567,6 +569,80 @@ function processWarehouseMessage(message, options) {
...
+      );
+      commonColumnTypes[eventColName] = 'string';
+
+      // -----start: event table------
```
Comment (achettyiitr) [inline src/warehouse/index.js:595]: Can we remove this?

File: src/warehouse/index.js [hunk 4]
```diff
@@ -567,6 +569,80 @@ function processWarehouseMessage(message, options) {
...
+
+      // do not create event table in case of empty event name (after utils.transformColumnName)
+      if (_.toString(commonProps[eventColName]).trim() === '') {
```
Comment (achettyiitr) [inline src/warehouse/index.js:598]: Should we need this check?

File: src/warehouse/index.js [hunk 5]
```diff
@@ -567,6 +569,80 @@ function processWarehouseMessage(message, options) {
...
+        'context_',
+      );
+
+      // set event column based on event_text in the tracks table
+      const eventColName = utils.safeColumnName(options, 'event');
+      commonProps[eventColName] = utils.transformTableName(
+        options,
+        message[utils.safeColumnName(options, 'event')],
+      );
+      commonColumnTypes[eventColName] = 'string';
```
Comment (achettyiitr) [inline src/warehouse/index.js:593]: Do we need this? Earlier we were doing this for tracks and events tables?

File: src/warehouse/index.js [hunk 6]
```diff
@@ -567,6 +569,80 @@ function processWarehouseMessage(message, options) {
...
+
+      // -----start: event table------
+
+      // do not create event table in case of empty event name (after utils.transformColumnName)
+      if (_.toString(commonProps[eventColName]).trim() === '') {
+        break;
+      }
```
Comment (achettyiitr) [inline src/warehouse/index.js:600]: Earlier we used this for tracks + events table compatibility. Do we need this for extract as well?

Comment (achettyiitr) [reply src/warehouse/index.js:600]: We should probably return error right incase of extract?

File: src/warehouse/index.js [hunk 7]
```diff
@@ -567,6 +569,80 @@ function processWarehouseMessage(message, options) {
...
+
+      setDataFromInputAndComputeColumnTypes(
+        utils,
+        eventType,
+        commonProps,
+        message.context,
+        commonColumnTypes,
+        options,
+        'context_',
+      );
```
Comment (achettyiitr) [inline src/warehouse/index.js:585]: Do we still need context?

File: src/warehouse/index.js [hunk 8]
```diff
@@ -66,6 +67,7 @@ const rudderReservedColums = {
   screen: { ...whDefaultColumnMappingRules, ...whScreenColumnMappingRules },
   group: { ...whDefaultColumnMappingRules, ...whGroupColumnMappingRules },
   alias: { ...whDefaultColumnMappingRules, ...whAliasColumnMappingRules },
+  extract: { ...whExtractEventTableColumnMappingRules }
```
Comment (achettyiitr) [inline src/warehouse/index.js:70]: Just wanted to confirm, we don't need these right?
```Go
const rules = {
  id: 'messageId',
  anonymous_id: 'anonymousId',
  user_id: 'userId',
  sent_at: 'sentAt',
  timestamp: 'timestamp',
  received_at: 'receivedAt',
  original_timestamp: 'originalTimestamp',
  channel: 'channel',
  context_ip: (message) => getFirstValidValue(message, ['context.ip', 'request_ip']),
  context_request_ip: 'request_ip',
  context_passed_ip: 'context.ip',
};
```

**Comment by achettyiitr** [comment]
@debanjan97 Can u write test cases as well?

File: src/warehouse/index.js (hunk 8 above)
Comment (debanjan97) [reply src/warehouse/index.js:70]: Yeah, we kept `id` and `receivedAt` for extract events

File: src/warehouse/index.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (debanjan97) [reply src/warehouse/index.js:585]: We decided to keep context columns as of now.
We have yet to decide on what columns can be skipped there, as in almost whatever field we send as part of the context, we rely on them while debugging.

File: src/warehouse/index.js (hunk 6 above)
Comment (debanjan97) [reply src/warehouse/index.js]: You are correct. I have raised an `InstrumentationError` in case the event name is missing from the payload.

**Review by achettyiitr — DISMISSED** [review]

**Review by achettyiitr — APPROVED** [review]

**Comment by debanjan97** [comment]
> @debanjan97 Can u write test cases as well?

Added test for the new event type

**Review by nidhilashkari17 — APPROVED** [review]

**Review by ItsSudip — APPROVED** [review]

**Pull request merged**